snippet_parser/core.py

```diff
@@ -191,94 +193,120 @@ def extract(self, wikitext):
             except:
                 continue
 
-            tree = lxml.html.parse(
-                StringIO.StringIO(html),
-                parser = lxml.html.HTMLParser(
-                    encoding = 'utf-8', remove_comments = True)).getroot()
-            if tree is None: continue
-
-            for strip_selector in self._html_css_selectors_to_strip:
-                for element in strip_selector(tree):
-                    # Make sure we don't remove elements inside the markers.
-                    # For a few Wikipedias (Chinese, Russian) the expansion of
-                    # {{fact}} is marked as .noprint, which we otherwise want
-                    # to remove.
-                    inside_marker = any(
-                        e.attrib.get('class') == CITATION_NEEDED_MARKER_CLASS
-                        for e in element.iterancestors('span'))
-                    if not inside_marker:
-                        lxml_utils.remove_element(element)
-
-            snippet_roots = []
-            if self._cfg.extract == 'snippet':
-                # We climb up from each marker to the nearest antecessor element
-                # that we can use as a snippet.
-                for marker in tree.cssselect('.' + CITATION_NEEDED_MARKER_CLASS):
-                    root = marker.getparent()
-                    while root is not None and root.tag not in _SNIPPET_ROOT_TAGS:
-                        root = root.getparent()
-                    if root is None:
-                        continue
-                    if root.tag in _LIST_TAGS:
-                        snippet_roots = self._html_list_to_snippets(root)
-                    else:
-                        snippet_roots = [self._make_snippet_root(root)]
-            else:
-                # Throw away the actual template, we don't need it.
-                for marker in tree.cssselect('.' + CITATION_NEEDED_MARKER_CLASS):
-                    lxml_utils.remove_element(marker)
-
-                # Keep only snippet root top-level elements within the body
-                # that have any text content (we may have created empty elements
-                # above during cleanup). This is not great as any content within,
-                # say, <blockquote> gets removed entirely, but it's good enough
-                # in most cases.
-                snippet_roots = [
-                    self._make_snippet_root(*(
-                        e for e in tree.cssselect(
-                            'body > ' + ', '.join(_SNIPPET_ROOT_TAGS))
-                        if e.text_content() and not e.text_content().isspace()))
-                ]
-
-            snippets_in_section = set()
-            for sr in snippet_roots:
-                # Some last-minute cleanup to shrink the snippet some more.
-                # Remove links and attributes, but make sure to keep the
-                # class in our marker elements, and that there is no space
```

Comment:
Oh, the "and that there is no space before it" might be the reason you were not seeing proper spacing in your tests.